### ISSUE TITLE
Microservice to talk to Airtable - push targets to campaign

### ISSUE THREAD

**Issue by marcinkoziej** [fixthestatusquo/proca-server#45]

**Comment by hiemanshu** [comment]
For Airtable, the only "official" library available is in Javascript. So creating a microservice makes sense using Node. The Node microservice will also be able to handle other services like Google Sheets.

Airtable, and Google Sheets will require minor configuration from Campaign Creators. In both cases:

1. Sharing with an email (or service ID given by Google) owned by Proca. 
2. Inputting in the ID of the Base or Sheet 

The Microservice in the first iteration will be a barebones scheduler app, and mostly communicate via GraphQL with Proca, or to Airtable and GoogleSheets using REST APIs.

### Microservice:

Simple microservice using typescript, and `node-cron` to run a cron to check for updated data. The microservice will not store any data by itself, it will only handle transforming data. Fetching `id` of bases / sheets from Proca, pulling in the data from Airtable, and ssending it back to proca formatted.

### Proca:

Proca will get 2 changes:

1. A new `campaign_data` table to store the data, for representatives, and also to store the service name, and service id to fetch the data from
2. Resolvers to handle this new data

**Comment by marcinkoziej** [comment]
1. Table with officials

Lets have a table `officials` (`Proca.Official`) that has a FK to campaign, in it:

```
uuid - the hard-to-guess uuid 
external_id - some uuid or other id that the microservice maintains to be able to update the proper records
enable: boolean
name
position
affiliation
area
language (?)
emails  {:array, :string} 
phone_numbers {:array, :string}
fax

```
Then we need some sort of official_events table to store data about their email not accepting emails?
Or maybe better: not `emails` column but have `official_emails` table for each email, with its status (so we can mark them as able to deliver or not).

2. Services are available for fetching using `org { services }` in GQL, the service has fields: host, path, user, password - so enough to configure access to Airtable.

3. The id/reference number/code of airtable document or google doc - it should be a column on `campaigns` (how to call it?) `address_book_ref`? (similar to email template reference - so whatever id the email backend uses)

4. Can the services get webhook triggers from the airtable/google doc when the documents are updated? Or not possible? Are we at risk of using up api call number limit if we will be polling them at regular interval?

5. uuid - generated so proca widget can list as targets of particular action. We don't want to expose email or other data to the widget, rather _uuid, name, position, affiliation_ for example, and after member does the select of who they want to contact, the widget sends uuid list as targets.

6. Updating - I imagine either:
  a. `upsertCampaingOfficials(name:, officials: list_of(:official))`  which would overwrite the officials list according to the external_id, or add the officials which are not found by external_id. XXX how to remove the official? perhaps only disabling is possible by setting enable: false
  b. `updateCampaignOfficials(name: officials: list_of(:official)` replaces the whole officials list, but this way I'm not sure how to keep associations. eg. `official_emails` table? perhps the `official_emails` should use an email and not officials.id as FK.

7. There needs to be some sort of API query for the microservice to get all the airtable document refs for all the campaigns in the system, Should it be accessibly by using `campaigns()` query to fetch all the campaigns?

**Comment by hiemanshu** [comment]
1. A separate official table also sounds good. Instead of `official_emails`, I'd recommend `official_contact` so we can have many types, emails, phone, etc if needed in the future, or is this a scenario that wouldn't happen?

2. The microservice will contain all the needed credentials. Which is just simple API keys for Airtable, and more. All the data will be fetched through a single account to which other users will have to share the airtable with.

3. `officials_list_ref` would work if we're calling the table of data `officials`.

4. The Services cannot yet send webhook triggers for this. We will need to set n8n or something similar for that, and we'll have to automate that, which is additional required work. I would suggest in Phase 1 just to do get the basics working, and then once that is done and tested, we can do it. Otherwise the feature will get too large to test effectively. For now I'd suggest updating the data only twice a day or less. Until we build the full integration with n8n. 

5. `Ecto.UUID` is a supported type that we can use for this. 

6. If we use `upsert` the microservice will have to handle knowing the previous IDs of the officials. Keeping it simple and justing `update` would be prefered, so the microservice can remain simple. We can have associations built using IDs as foreign keys, since the old records will get deleted and new ones will be created with new relations. 

7. Yup, using an simple query with filter `campaigns(filter: { has: officials_list_ref })` will return all campaigns with a valid ref id

8. For now, the microservice will just the push the data using a mutation. Once the MVP is done, we can setup n8n, and automate it, to have the microservice only be activated on new record updated.

**Comment by marcinkoziej** [comment]
In reference to the other issue, let's s/official/target/g so targets, target_emails

**Comment by marcinkoziej** [comment]
Ad 6: of i have a target_email record flagged as bouncing and then I do an update like this how will I reconnect it to new records? I thought you send some external_id to match on

**Comment by hiemanshu** [comment]
Yup, `targets` works

6: Hmm yeah. Makes sense. Let the microservice store that information then. So `upsert` it is

**Comment by tttp** [comment]
@hiemanshu + @marcinkoziej we need two outputs for the target lists:
- the full data (including email or phone number) pushed to the server
- the front end data (used by the widget) that only contains public information (eg the name and country, but not the email or phone)

the widget would say "send this message to target in {key1}, {key2}" (graphql api)
the server +sending service would need to retrieve the private data (the email) from the key
 
# key
So we need some common key between the two. is airtable able to provide that or shall we assume we might need to do it?
a common key we used in past campaign is the twitter id. It's not hard to get, but does it need to be @marcinkoziej ?

# the widget 

# On the microservice-cron: 
I don't think it's a good idea to cron it, updating the list of target should be an intentional request from the campaign coordinators (big button "update" in proca-admin)

# On the microservice:
it seems to be possible to use n8n to do the cron/webhook + transformation + generating json + pushing to proca
Could you try? Performance isn't an issue, and it would allow to modify easily the transform function 

# on the table struture
@marcinkoziej , what's the point of storing as separate fields position area and so on? 
the widget will only say send the message to key1, the rest *might* be useful for the sending, but doesn't need to be structured
eg what about a key, name (for convenience) + details (json blob)
what's the benefit of having emails and so on out of the details and into separate columns?

**Comment by tttp** [comment]
> Ad 6: of i have a target_email record flagged as bouncing and then I do an update like this how will I reconnect it to new records? I thought you send some external_id to match on

I'm not sure on that one: if we start modifying a record mid campaign, is it ok to say we sent 200k to the PM of france, even so it was 100k to primeminister@france.fr and 100k at press@matignon.fr ?

I think we need some tracability here, if we upsert the record, we will loose the ability to tell who (which email) it was sent to. 

May be it's fine to upsert and we need to update each action with the information about who they got the email sent to? may be we should add two timestamps (use from, use to), to handle the modification?

**Comment by hiemanshu** [comment]
> @hiemanshu + @marcinkoziej we need two outputs for the target lists:
> 
>     * the full data (including email or phone number) pushed to the server
> 
>     * the front end data (used by the widget) that only contains public information (eg the name and country, but not the email or phone)

Correct. This is how it's planned to handle. Data will be read from Airtable and full data will be sent over.

> the widget would say "send this message to target in {key1}, {key2}" (graphql api) the server +sending service would need to retrieve the private data (the email) from the key

Correct. The only public data will be a key, and their name possibly. Private information like email / phone will not be shown.

> # key
> So we need some common key between the two. is airtable able to provide that or shall we assume we might need to do it? a common key we used in past campaign is the twitter id. It's not hard to get, but does it need to be @marcinkoziej ?

Airtable does have record IDs. The plan was to create a record to UUID mapping to handle the conversion from Airtable data to proca server data. So the UUID created in Proca will hold the main ID / key.

> # On the microservice-cron:
> 
> I don't think it's a good idea to cron it, updating the list of target should be an intentional request from the campaign coordinators (big button "update" in proca-admin)

I want to add this as a super temporary stop gap to have a complete feature. The section below covers more details on this.

> # On the microservice:
> 
> it seems to be possible to use n8n to do the cron/webhook + transformation + generating json + pushing to proca Could you try? Performance isn't an issue, and it would allow to modify easily the transform function

Yeah, it's possible to use n8n to do the "If record is updated, notify microservice", and then the microservice will decide what to do with that information. The plan is to split this into 2 features, because having all of this be 1 ticket is too large a ticket, and a lot of work and setup to be done in one go, and to be tested correctly, so initially for internal testing only, a cron will be setup to make sure things like upserting data, etc are handled properly. The amount of work to setup n8n is going to be a little bit since it will need the microservice to setup to handle n8n configurations, etc, so that it can happen automatically, and doesn't require someone to set it up manually. n8n has an internal undocumented API that can be used to set this up. 

> # on the table struture
> @marcinkoziej , what's the point of storing as separate fields position area and so on? the widget will only say send the message to key1, the rest _might_ be useful for the sending, but doesn't need to be structured eg what about a key, name (for convenience) + details (json blob) what's the benefit of having emails and so on out of the details and into separate columns?

I think splitting data up into multiple tables will allow for easy indexing of data, and allow for more generic data storage. For now we might start with just emails, but for future we might need to store more data, so my suggestion for this is to use a `target_contact` table that will contain a FK to the target, type of contact, and value. This will allow more structured searchable data with less read overhead.

**Comment by marcinkoziej** [comment]
@tttp Regarding the key:
> # key
> 
> So we need some common key between the two. is airtable able to provide that or shall we assume we might need to do it?
> a common key we used in past campaign is the twitter id. It's not hard to get, but does it need to be @marcinkoziej ?

We generate that uuid in proca server, and then it gets exported to the static target cache of the widget.

**Comment by marcinkoziej** [comment]
> I don't think it's a good idea to cron it, updating the list of target should be an intentional request from the campaign  coordinators (big button "update" in proca-admin)

I get you want to avoid the update-by-mistake problem here, but this way they have to log in to airtable and also to proca dashboard,... we are creating a kind of UX annoyance for them, just to "keep them safe". maybe we should rather inform them that updates on airtable WILL get to the widget and _this is not a drill_

**Comment by marcinkoziej** [comment]
>  what's the benefit of having emails and so on out of the details and into separate columns? 

- to have primary, secondary email
- to be able to mark metrics for each email (how many sent, how many opened, clicked and so on), does it bounce?

**Comment by marcinkoziej** [comment]
> what's the point of storing as separate fields position area and so on?
> the widget will only say send the message to key1, the rest might be useful for the sending, but doesn't need to be structured
> eg what about a key, name (for convenience) + details (json blob)

I would keep the searchable, standard data in columns, and some particular custom-for-campaign fields in details.
Area is also a column in supporters, the particular contact data type can define what area is used there, and we could match that to retrieve the targets for such area...
The position or affiliation is indeed not very interesting to proca, so could be in details. I would always think whether we would want to filter on it or not

**Comment by marcinkoziej** [comment]
> I'm not sure on that one: if we start modifying a record mid campaign, is it ok to say we sent 200k to the PM of france, even > so it was 100k to primeminister@france.fr and 100k at press@matignon.fr ?

^^ all these questions miss the point of having the emails stored in `target_emails` not `targets`

**Comment by marcinkoziej** [comment]
> target_contacts vs target_emails

I think we should resist the temptation to generalize from unknown cases. Now we have MTT and we could have Click to call soon, but I guess both would have different columns (emails can bounce, and phone numbers? do they bounce in the same way?). 
I'd rather do `target_emails` now, and focus on this, and when we will be adding the second kind of contact data, we will run a migration to rename it to target_contacs if it makes sense. If not, we will add target_phones

**Comment by hiemanshu** [comment]
Cool. Sounds good.

**Comment by tttp** [comment]
we went for n8n
